src/lib/storage/shadowed_vol_helper.rb

```diff
@@ -0,0 +1,152 @@
+    # not shadowed anymore
+    #
+    # In order to make the recovery of previously deleted subvolumes possible,
+    # it stores the deleted volumes in a class variable. This list needs to be
+    # cleaned between subsequent calls to the expert partitioner.
+    # @see .cleanup
+    def root_partition
+      log.info " shadowed subvolumes"
+      return unless doable?
+      root_copy = deep_dup(root_part)
+      restore_aborted_subvols(root_copy)
+      abort_shadowed_subvols(root_copy)
+      log.info "New list of subvolumes: #{root_copy["subvol"]}"
+      root_copy
+    end
+
+    # Cleans the global list of deleted subvolumes
+    def self.cleanup
+      clean_aborted_subvols
+    end
+
+  protected
+
+    def deep_dup(hash)
+      Marshal.load(Marshal.dump(hash))
+    end
+
+    def partitions
+      target_map.values.each_with_object([]) do |disk, list|
+        list.concat(disk.fetch("partitions", []))
+      end
+    end
+
+    def root_part
+      partitions.detect {|p| p["mount"] == "/" }
+    end
+
+    def doable?
+      if root_part.nil?
+        log.info "No root partition found, skipping shadowed volumes update"
+        false
+      else
+        true
+      end
+    end
+
+    def abort_shadowed_subvols(target_part)
+      partitions.each do |part|
+        if part != root_part && part.has_key?("mount")
+          abort_shadowed_subvols_for(target_part, part["mount"])
+        end
+      end
+    end
+
+    # Aborts creation of shadowed subvolumes
+    def abort_shadowed_subvols_for(target_part, mount_point)
+      shadowed = ShadowedVolList.new(partition: target_part, mount_point: mount_point)
+      shadowed.each do |subvol|
+        log.debug "Subvol to abort: #{subvol}"
+        # Only abort planned volumes (don't delete pre-existing ones)
+        if !subvol["create"]
+          log.info "Skipping subvolume #{subvol["name"]}"
+          next
+        end
+
+        log.info "Deleting subvolume #{subvol["name"]}"
+        # Store it, so we can restore it
+        memorize(subvol)
+        # Mark it for removal ("delete" takes precedence over "create")
+        subvol["delete"] = true
+      end
+      log.debug "Aborted subvols: #{aborted_subvols}"
+    end
+
+    def memorize(subvol)
+      aborted_subvols << subvol.dup
+    end
+
+    def restore_aborted_subvols(target_part)
+      aborted_subvols.each do |subvol|
+        name = subvol["name"]
+        log.info "Restore subvol #{name}?"
+        log.debug "subvol: #{subvol}"
+        existing_subvol = target_part["subvol"].detect { |s| s["name"] == name}
+        if existing_subvol.nil?
+          log.info "Restoring subvol"
+          target_part["subvol"] << subvol
+        elsif existing_subvol["delete"]
+          log.info "Rejecting subvol deletion"
+          existing_subvol["delete"] = false
+        end
+      end
+      clean_aborted_subvols
+    end
+
+    def aborted_subvols
+      @@aborted_subvols ||= []
```

Comment:
I use a singleton instance for keeping persistent data (https://github.com/yast/yast-registration/blob/master/src/lib/registration/storage.rb#L31), IMHO it's a bit more consistent with "normal" objects. But I'm not against class attributes, I simply do not know which one is better... :grey_question: 

But we should be consistent across YaST modules and use one pattern preferably... a topic for discussion?